### ISSUE TITLE
Provide binaries for windows

### ISSUE THREAD

**Issue by ddobrev** [semlanik/qtprotobuf#188]
**Problem description**
Building qtprotobuf is quite difficult outside of Linux because of the numerous dependencies. macOS poses trouble because not everybody owns a Mac but binaries even just for windows will help immensely.

**Suggestion/solution**
Use a virtual machine or use CI (such as Actions here) to produce and upload a binary package for windows 7 and above.

**Alternatives**
Building qtprotobuf from source code which is obviously difficult without even having to try just looking at the sheer number of dependencies.

**Comment by gmabey** [comment]
Basic functioning binaries for windows (just protobuf, not `grpc`) would be easier after #181 were addressed, just a thought.

**Comment by semlanik** [comment]
It could be complete in CI, I think, but where is "travis"?! Need to find out what happened with the windows sanity check.

**Comment by ddobrev** [comment]
@semlanik thank you very much for taking care.

**Comment by ddobrev** [comment]
Would you please consider adding this to the 0.6.0 milestone?

**Comment by semlanik** [comment]
@ddobrev I will try. The only question is when I will be able to release 0.6? I already crossed the deadline, I guess 5 times, so hopefully, I will finish planned features this month.

**Comment by ddobrev** [comment]
@semlanik thank you for your consideration. I would definitely prefer not to slow you down, so we could perhaps discuss another solution. What do you think about this: since I understand that your win build itself works, the major amount of time would go into writing the scripts for CI. You could, however, do it later, and simply pack a zip archive and upload in our releases. User, including myself, would then have convenient binaries to test with while you continue your proper support for the platform.

**Comment by ddobrev** [comment]
@semlanik so, what do you think about this idea? "Unofficial" binaries for windows so that your release isn't put off any more?

**Comment by semlanik** [comment]
Currently, I could only provide QtProtobuf, the same way as it's already done for the Linux platforms. So you need build/find the protoc executable manually. Distributing protoc and side libraries sounds quite difficult for now.

**Comment by semlanik** [comment]
Hi @ddobrev , please look at the resulting zip package here:
    https://github.com/semlanik/qtprotobuf/releases/download/refs%2Fheads%2Fci_check/qtprotobuf-0.5.0.zip
A similar package will be provided for the 6.0 release.

**Comment by ddobrev** [comment]
@semlanik looks absolutely stunning!
Does it need anything else to work? Such as Protobuf itself, as you said earlier? If so, do you have a list of the requirements? Is it just the executable, or the library too?

**Comment by ddobrev** [comment]
In addition, I don't see `microjson`, is it statically stuffed in one of the `qtprotobuf` libs?

**Comment by semlanik** [comment]
Yeah, microjson is linked statically as an object library for both the static and shared libraries.

**Comment by semlanik** [comment]
> @semlanik looks absolutely stunning!
> Does it need anything else to work? Such as Protobuf itself, as you said earlier? If so, do you have a list of the requirements? Is it just the executable, or the library too?

It needs a whole set of the protobuf libraries and protoc, as I mentioned before. But I will change this soon, so at runtime only protoc will be required.

**Comment by ddobrev** [comment]
protoc at run-time? But it only generates the code so why would I need it at run-time?

**Comment by semlanik** [comment]
at generator runtime. When I mean run-time it's mainly about the QtProtobuf runtime but not about the user application runtime. :)

**Comment by ddobrev** [comment]
@semlanik I'd be happy to know a little more about the way qtprotobuf works. :) Why does it need `protoc`? In theory it should only require the `*.proto`-s as input and produce the `QObject`-s as output.

**Comment by semlanik** [comment]
QtProtobuf only implements the generator plugin for protoc. protoc is a standalone parser implemented by google and it has a pretty clear interface, so I didn't and don't see the need of implementing my own proto parser. So that's why to generate QObject's from the .proto files it's required to have protoc "at runtime".

QtProtobuf meanwhile doesn't link libprotobuf, so userspace applications don't need to depend on reference protobuf. The only exceptions is WellKnowTypes library that requires the .proto files of libprotobuf to specify the correct proto include paths for the protoc call, at user application compile time.

The situation with the gRPC library is almost the same. The exception is the native gRPC transport implemented by one of the contributors. It has a binary runtime dependency on grpc++ as far as I remember. I will decompose this part soon.

**Comment by ddobrev** [comment]
So Protobuf is extensible and you've simply implemented an extension. The smart way.
Thank you for your explanation. I like your work more and more. I'm just waiting for https://github.com/semlanik/qtprotobuf/issues/195 and `0.6.0` and I'll be good to try integrating this marvellous piece of software in our project.
